### ISSUE TITLE
Too many connections errors

### ISSUE THREAD

**Issue by vtoulemonde** [kenjis/ci-phpunit-test#275]
In a controller test class, after the execution of around 20-30 test functions, I have a db error:

```
Message:     mysqli::real_connect(): (HY000/1040): Too many connections
Filename:    ..... app/system/database/drivers/mysqli/mysqli_driver.php
Line Number: 201
```

Test functions are really basic for now and only do the request and 1 assertion:

```php
class Contracts_test extends TestCase {

	public function setUp() {
		$this->resetInstance();
		$this->CI->tank_auth->login(XXXXXXX);

	}

	public function test_dashboard() {
		$output = $this->request('GET', 'contracts/dashboard');
		$this->assertContains('<title>XXXXXXX</title>', $output);

	}

}
```

**Comment by kenjis** [comment]
`$this->request()` resets CodeIgniter instance. So it seems your `setUp()` does not work.
What if you remove the `setUp()`?

**Comment by vtoulemonde** [comment]
The code in `setUp()` is to load my user in the session so that when I access controller functions with `$this->request()`  the user is well logged-in.
Same error happens when I remove `setUp()`.

**Comment by kenjis** [comment]
Every time when you call `$this->request()` or `$this->resetInstance()`, a new db connection is probably created. So if the connection does not close, you will reach the max connections of MySQL.

When a db instance is destructed (by Garbage Collection), the db connection also close.

You get the too many connections error, so you probably have many db instances (not garbage collected).

**Comment by vtoulemonde** [comment]
So how do I force the db connection to close?

**Comment by ytetsuro** [comment]
@kenjis 

> When a db instance is destructed (by Garbage Collection), the db connection also close.

I think not garbage collected.
the that reason, please see following.

https://phpunit.readthedocs.io/en/latest/fixtures.html#fixtures-more-setup-than-teardown
> The garbage collection of test case objects is not predictable.

I think that problem encountering when many test case with db connection are execute.
What do you think this pr solution?

https://github.com/kenjis/ci-phpunit-test/pull/276

**Comment by vtoulemonde** [comment]
Thanks for your solution! It did not work as is because the errors happen
during the execution of one test class with a lot of request calls and the
tearDownAfterClass() is executing after the errors happen. But it is
working if I put the code in tearDown() function. Nice!

**Comment by ytetsuro** [comment]
Thanks for look my PR!
> It did not work as is because the errors happen

Hmm...

I think occurs error when one test class has many database connection.
but, occurs problem when the developer writes the following such a test code.

```php
<?php

class Example_test extends TestCase
{
   private static $db;

   public static function setUpBeforeClass()
   {
      parent:: setUpBeforeClass();
      $CI =& get_instance();
      self::$db = $CI->load->database('unit-test', true);
   }

   public function test_when_first()
   {
      $row = self::$db->get('user', 1);
      $this->assertSame($row->id, '1');
   }

   public function test_when_second()
   {
      $row = self::$db->get('user', 1); // gone when close db connection from tearDown.
      $this->assertSame($row->id, '1');
   }
}
```

Do you think this such a test code is rare case?
Should I change the PR to close the database with tearDown?

**Comment by vtoulemonde** [comment]
For testing controllers (my case), I think the issue is that the connections are not closed after calling $this->request(). Calling request() will create a new CI instance, so a new db connection, but the potential previous db connections are not closed then. So that's something that can be done at the end of the request(), and in the resetInstance().
When testing models and libraries, the developer might want to use the same db instance for a functions in the same class so releasing connections in the tearDown() will break that.
To handle all those cases I would say to release connections at the end of request(), at the beginning of resetInstance() and in the tearDownAfterClass().

**Comment by kenjis** [comment]
@vtoulemonde @alexfraundorf-com 
I merged #276 into master branch.
If there is something wrong, please inform me.

**Comment by ytetsuro** [comment]
I have a similar error when using ORACLE, but when I apply #276, the error no longer occurs.
I wrote this comment for someone's search.
`ORA-01000: maximum open cursors exceeded`
